oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2DeviceCodeAuthenticationProvider.java

```diff
@@ -0,0 +1,245 @@
+		RegisteredClient registeredClient = clientPrincipal.getRegisteredClient();
+		Assert.notNull(registeredClient, "registeredClient cannot be null");
+
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Retrieved registered client");
+		}
+
+		OAuth2Authorization authorization = this.authorizationService.findByToken(
+				deviceCodeAuthentication.getDeviceCode(), OAuth2DeviceCode.TOKEN_TYPE);
+		if (authorization == null) {
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+		}
+
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Retrieved authorization with device code");
+		}
+
+		OAuth2AuthorizationRequest authorizationRequest = authorization.getAttribute(
+				OAuth2AuthorizationRequest.class.getName());
+		Assert.notNull(authorizationRequest, "authorizationRequest cannot be null");
+
+		OAuth2Authorization.Token<OAuth2DeviceCode> deviceCode = authorization.getToken(OAuth2DeviceCode.class);
+		Assert.notNull(deviceCode, "deviceCode cannot be null");
+
+		if (!registeredClient.getClientId().equals(authorizationRequest.getClientId())) {
+			if (!deviceCode.isInvalidated()) {
+				// Invalidate the device code given that a different client is attempting to use it
+				authorization = OAuth2AuthenticationProviderUtils.invalidate(authorization, deviceCode.getToken());
+				this.authorizationService.save(authorization);
+				if (this.logger.isWarnEnabled()) {
+					this.logger.warn(LogMessage.format(
+							"Invalidated device code used by registered client '%s'", registeredClient.getId()));
+				}
+			}
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+		}
+
+		// In https://www.rfc-editor.org/rfc/rfc8628.html#section-3.5,
+		// the following error codes are defined:
+
+		//   access_denied
+		//      The authorization request was denied.
+		if (Boolean.TRUE.equals(deviceCode.getMetadata(OAuth2Authorization.Token.ACCESS_DENIED_METADATA_NAME))) {
+			OAuth2Error error = new OAuth2Error("access_denied", null, DEVICE_ERROR_URI);
+			throw new OAuth2AuthenticationException(error);
+		}
+
+		//   expired_token
+		//      The "device_code" has expired, and the device authorization
+		//      session has concluded.  The client MAY commence a new device
+		//      authorization request but SHOULD wait for user interaction before
+		//      restarting to avoid unnecessary polling.
+		if (deviceCode.isExpired()) {
+			OAuth2Error error = new OAuth2Error("expired_token", null, DEVICE_ERROR_URI);
+			throw new OAuth2AuthenticationException(error);
+		}
+
+		//   slow_down
+		//      A variant of "authorization_pending", the authorization request is
+		//      still pending and polling should continue, but the interval MUST
+		//      be increased by 5 seconds for this and all subsequent requests.
+		if (deviceCode.isBeforeUse()) {
+			OAuth2Error error = new OAuth2Error("slow_down", null, DEVICE_ERROR_URI);
```

Comment:
My first thought was that the framework probably wouldn't have a way to determine when this error is appropriate and so wouldn't return it at all. But if the token is customized to contain an `nbf` claim and the client called the token endpoint before that time, it could be a clear case that this error applies. However, to your point it's probably not the right signal that the error is needed. We can discuss this one.

Reply:
We will remove `slow_down` entirely for now.